src/Nethermind/Nethermind.Core/Threading/ThreadLimiter.cs

```diff
@@ -0,0 +1,46 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Threading;
+
+namespace Nethermind.Core.Threading;
+
+/// <summary>
+/// Encapsulate the pattern of checking if new task can be spawned based on a predefined limit.
+/// Used in multithreaded tree visit where we don't know if we can spawn task or not and spawning task itself
+/// is not a cheap operation.
+///
+/// Yes, I don't like the name. Give me a good one.
+/// </summary>
+/// <param name="concurrency">Desired concurrency which include the calling thread. So slot is slot-1.</param>
+public class ThreadLimiter(int concurrency)
+{
+    private int _slots = concurrency - 1;
+
+    public bool TryTakeSlot(out SlotReturner returner)
+    {
+        returner = new SlotReturner(this);
+        int newSlot = Interlocked.Decrement(ref _slots);
+        if (newSlot < 0)
+        {
+            Interlocked.Increment(ref _slots);
```

Comment:
Can this be done better with `voliatale`? @Scooletz

Reply:
It can't as you want to undo the previous `Decrement` in an atomic way.